### PR TITLE
feat: add poddefault for automatic Jupyterlab authentication

### PR DESCRIPTION
**Pull request by juliusvonkohout**
**Description of your changes:**
@Bobgy 
A poddefault is automatically created according to https://github.com/kubeflow/pipelines/issues/5138

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by thesuperzapper** [comment]
@Bobgy this is probably something we shouldn't decide for users, (not all admins will want to auto-mount a token).

---

Also, I think this PR highlights that we should extend the main `Profile Controller` to remove the need for the hacky [`pipeline-profile-controller` python script](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py). For example, we can make a new `ProfileResourceTemplate` CRD, which defines arbitrary K8S templates to be automatically added to each Profile (by the profile controller).

Then, the Kubeflow Pipelines YAML would only need to include those `ProfileResourceTemplate` resources.

**Comment by juliusvonkohout** [comment]
> @Bobgy this is probably something we shouldn't decide for users, (not all admins will want to auto-mount a token).
> 
> Also, I think this PR highlights that we should extend the main `Profile Controller` to remove the need for the hacky [`pipeline-profile-controller` python script](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py). For example, we can make a new `ProfileResourceTemplate` CRD, which defines arbitrary K8S templates to be automatically added to each Profile (by the profile controller).
> 
> Then, the Kubeflow Pipelines YAML would only need to include those `ProfileResourceTemplate` resources.

It is not auto mounted. it is just providing the poddefault that might/can be used in a Jupyterlab. https://github.com/kubeflow/kubeflow/pull/6160 would auto mount it by default. But if you have a cleaner solution that is of course appreciated.

**Comment by juliusvonkohout** [comment]
Is the profile controller able to monitor these resources and recreate them if they have been deleted?

**Comment by ca-scribner** [comment]
Yes, I believe the profile controller repeatedly does a reconciliation loop (every 10s maybe?  something like that) and if it ever sees a missing resource, it creates a new one.  The sync.py script does not do really detailed inspection of the objects though, so for example if you edited the proposed PodDefault and just deleted the spec, the profile controller wouldn't notice that or fix anything

**Comment by ca-scribner** [comment]
@thesuperzapper I don't know if admins will want to auto-mount tokens, but this feature feels like an easy quality of life improvement for users that doesn't change much for admins.  It creates an object that users have permission to create for themselves (any user could create this `PodDefault` in their namespace), it just means they don't need to know how.

Having seamless access to the pipeline client from notebooks feels like a feature most users would expect.  I really like the `ProfileResourceTemplate` CRD idea and would love to have that.  In the interim though, this feels like a nice low-cost fix.

File: manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py [hunk 1]
```diff
@@ -339,6 +340,51 @@ def sync(self, parent, children):
                         }
                     }
                 },
+                {
```
Comment (ca-scribner) [inline manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py:343]: Suggest adding a header so this doesn't look like part of the above artifact fetcher resources (I'm not particular about the wording, just don't want it to read that this is part of the artifact server)
```suggestion
                # Adds "Allow access to Kubeflow Pipelines" button in Notebook spawner UI
                {
```

**Comment by ca-scribner** [comment]
If there is agreement on making this change, the tests in `pipelines/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/test_sync.py` should be updated

**Comment by juliusvonkohout** [comment]
> If there is agreement on making this change, the tests in `pipelines/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/test_sync.py` should be updated

Actually i and @thesuperzapper want to get rid of metacontroller and pipelines-profile-controller as described in https://github.com/kubeflow/pipelines/pull/7219#issuecomment-1024086393. This WIP pull request also contains the change you proposed above. Are you willing to help there?

**Comment by ca-scribner** [comment]
I am for removing the profile controller and metacontroller, and really like the `ProfileResourceTemplate` idea (@thesuperzapper have you opened that as a separate feature issue?  I didn't see one).  

Re #7219 I like what you've outlined (adding comment there too).  It would be good to get someone from the pipelines group to express buy-in.  I cannot contribute to either of these immediately, but can probably help within a few weeks.

**Comment by thesuperzapper** [comment]
@kimwnasptd the pipeline docs I added a while ago show an example of using a PodDefault to automatically mount the KFP ServiceAccount token and set `KF_PIPELINES_SA_TOKEN_PATH`.

Expand the ["Full Kubeflow (from inside cluster)"](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub) section

**Comment by juliusvonkohout** [comment]
closed due to inactivity

**Pull request closed without merging**